iocore/net/quic/QUICHandshake.cc

```diff
@@ -462,11 +459,16 @@ QUICHandshake::do_handshake()
   out.buf                                = out_buf;
   out.max_buf_len                        = MAX_HANDSHAKE_MSG_LEN;
 
-  int result = this->_hs_protocol->handshake(&out, &in);
-  if (this->_remote_transport_parameters == nullptr) {
-    if (!this->check_remote_transport_parameters()) {
-      result = 0;
+  int result = 0;
+  if (!this->is_completed()) {
+    result = this->_hs_protocol->handshake(&out, &in);
+    if (this->_remote_transport_parameters == nullptr) {
+      if (!this->check_remote_transport_parameters()) {
+        result = 0;
+      }
     }
+  } else {
+    result = this->_hs_protocol->read_ex(&out, &in);
```

Comment:
I'm bit confused. `NewSessionTicket` is in `out`, right? Do we need to send it back?

Reply:
I'm not sure your point. 

In my test 

client                                    server

CH ----------------------> 

    <----------------------- SH

FINISHED ----------------> 

      <---------------------- New Sesssion Ticket 

>I'm bit confused. NewSessionTicket is in out, right? Do we need to send it back?

Actually `NewSessionTicket ` is in `in`.

`read_ex` only be call in client side so it does not generate new session ticket msg(Actually it generate nothing).  And we also write something to the specified stream only if `_hs_protocol` write something in `out`. We will still send it to server.

Reply:
> Sounds like we don't need to pass out to read_ex.

I think I'm wrong. We need to pass `out` for some other TLS extensions.

Reply:
:-)

Reply:
It seems like what we want to do here is processing New Session Ticket message (To make OpenSSL to read the message and call a callback function). If so, I think we should not add `read_ex` to `QUICHandshake`.

I assume the name came from `SSL_read_ex` but the main purpose of `SSL_read_ex` is reading application data on a connection. It would process some TLS messages under the hood and it would occur some callbacks, but I'd call it a side effect. Since QUIC doesn't read / write application data on streams for handshake, adding `read_ex` to QUICHandshake doesn't make sense and this is a reason we are confused I think.

I'm ok with calling `SSL_read_ex` if there is no alternative way, but that is just an internal implementation. BoringSSL or other TLS implementation may have different functions for it. I don't want to expose OpenSSL specific function names.